lib/transport/logtransport.c

```diff
@@ -27,6 +27,65 @@
 
 #include <unistd.h>
 
+/*
+ * log_transport_read_chunk():
+ *
+ * Reads up to count bytes even if we need multiple read invocations.
+ *   * it attempts to read as many bytes as requested (e.g. count)
+ *
+ *   * if we encounter an error during reading of the input, that error is
+ *     returned, potentially discarding partial data.
+ *
+ *   * if we encounter EOF we return partial data (or EOF it we don't have
+ *     partial data)
+ *
+ *   * if we encounter EAGAIN, we return partial data (e.g.  this is not a
+ *     busy loop)
+ *
+ * Returns: the number of bytes read
+ */
+gssize
+log_transport_read_chunk(LogTransport *self, gpointer buf, gsize count, LogTransportAuxData *aux)
```

Comment:
This is implemented so as the TCP layer does not guarantee that even if messages are sent as single packets, then read() will also return them whole. If there's an active proxy sitting between the client and the sender we might get different fragmentation.

With that said, the current code will happily retry the read operation, until enough bytes are accumulated, so actually the loop may not add too much value here (except maybe avoiding going back to the main loop). I vaguely remember not wanting to do the retry mechanism, just as recommended by the RFC, but at the end it was too flimsy, so I added it anyway.

All that said, a read_chunk() function is more useful in blocking mode, where EAGAIN will never be received, but multiple read operations might be needed for a specific number of bytes to be read. In case an fd is in non-blocking mode, it behaves basically the same as read() except the short read case is not returned unless there really is not enough bytes.

It could probably be dropped, I just grew some fondness having it implemented.

Reply:
Thank you.